### DISCUSSION TITLE
Invalid input syntax for type integer

### DISCUSSION THREAD

**Discussion by jordan-carson** [psycopg/psycopg#668] [Q&A] [answered]
Currently migrating a few ETL pipelines to psycopg 3.1.12 - `psycopg = {extras = ["binary"], version = "^3.1.12"}`

When doing so I am using the cursor.copy() method to write csv files to postgres. My script below does this, however it breaks due to a variable being read as an integer rather than a string.

```python
file_name = r"/tmp/out/names.csv"

create_table = """
    CREATE TABLE IF NOT EXISTS main.names_example
    (
        id SERIAL PRIMARY KEY,
        names_id VARCHAR,
        name VARCHAR,
        companies VARCHAR
    );
"""


with psycopg.connect(create_conn_str("php")) as conn:
    with open(file_name, mode="r", encoding="utf8") as file:
        with conn.cursor() as curr:
            curr.execute(create_table)
            with curr.copy("COPY main.names_example FROM STDIN (format csv, delimiter '|', quote '\"')") as copy:
                next(file, None)    # Skipping headers...
                copy.write(file.read())

```

This produces the following error.

```
psycopg.errors.InvalidTextRepresentation: invalid input syntax for type integer: "N12892330"
CONTEXT:  COPY names_example, line 1, column id: "N12892330"
```

Is there a way to provide data types during the write?

**Comment by dvarrazzo** [comment, marked as answer]
`N12892330` is not an integer, it seems pretty unambiguous.

Maybe the order of the columns in the csv and the ones in the database don't match. You should specify the column list.

Otherwise yes, you can do some transformation, but you should copy row-by-row, not block-by-block.

**Reply by jordan-carson** [reply]
Where would one specify the column list?

**Reply by dvarrazzo** [reply]
See the [COPY documentation](https://www.postgresql.org/docs/current/sql-copy.html).
